lib/puppet/parser/resource/param.rb

```diff
@@ -19,4 +19,9 @@ def line_to_i
   def to_s
     "#{self.name} => #{self.value}"
   end
+
+  def self.from_param(param, value)
+    new(
+      :name => param.name, :value => value, :source => param.source)
```

Comment:
I think we need to copy additional fields from `param` such as `file`, `line`, etc